espnet2/asr/transducer/beam_search_transducer.py

```diff
@@ -244,18 +261,58 @@ def default_beam_search(self, enc_out: torch.Tensor) -> List[Hypothesis]:
             nbest_hyps: N-best hypothesis.
 
         """
+        logging.info("decoder input length: " + str(enc_out.shape[0]))
         beam = min(self.beam_size, self.vocab_size)
         beam_k = min(beam, (self.vocab_size - 1))
 
         dec_state = self.decoder.init_state(1)
 
         kept_hyps = [Hypothesis(score=0.0, yseq=[self.blank_id], dec_state=dec_state)]
         cache = {}
+        cache_lm = {}
+
+        def make_lm_tokens(yseq: List[int]) -> torch.Tensor:
+            """Make LM tokens from a list of tokens
+
+            If the first token is <blank>, then replace it with <sos>.
+            Return token sequence tensor for LM scoring.
+
+            """
+            if len(yseq) and yseq[0] == self.blank_id:
+                return torch.LongTensor(
+                    [self.sos] + yseq[1:], device=self.decoder.device
+                )
+            else:
+                return torch.LongTensor(yseq, device=self.decoder.device)
+
+        def clear_lm_cache(hyps):
+            """Clear LM cache
+
+            Keep cache for sequences in input hyps and clear outdated hypothesis.
+
+            """
+            keep_yseq = set([tuple(hyp.yseq) for hyp in hyps])
+            dump_yseq = [yseq for yseq in cache_lm if yseq not in keep_yseq]
+            for yseq in dump_yseq:
+                cache_lm.pop(yseq)
 
         for enc_out_t in enc_out:
+            clear_lm_cache(kept_hyps)
```

Comment:
How does the lm cache grows? Is it a bottleneck to keep them all (just asking)?

Reply:
At the time I wrote this, it felt natural to me to clear the cache to avoid increasing the dictionary size. I now checked the memory usage for some samples with very long text sequences. The maximum memory used by a single CPU without clearing cache was 1305MB and with clearing cache was 1270MB. Since there's no bottleneck, there's no reason to clear cache.  I will remove this function. Thank you for your careful review and pointing out all the above changes. With these modifications, there should only be a few changes overall.